### PR TITLE
#283: Add support for command EXPIREAT

### PR DESCRIPTION
**Pull request by ashwaniYDV**
PR to address: Support for command EXPIREAT
Link to https://github.com/DiceDB/dice/issues/283

EXPIREAT sets a expiry time(in unix-time-seconds) on the specified key in args
args should contain 2 values, key and the expiry time to be set for the key
The expiry time should be in integer format; if not, it returns encoded error response
Returns RespOne if expiry was set on the key successfully.
Once the time is lapsed, the key will be deleted automatically

### CODE REVIEW

**Comment by JyotinderSingh** [comment]
Please resolve the conflicts, will review once done.

**Comment by JyotinderSingh** [comment]
Just gave the PR a skim, please add more comprehensive unit and integration tests with edge case being covered.

**Comment by ashwaniYDV** [comment]
@JyotinderSingh I have added unit tests and integration tests

**Comment by ashwaniYDV** [comment]
Thanks for the reviews @JyotinderSingh 
I am adding unit tests

File: core/commands.go [hunk 1]
```diff
@@ -570,6 +570,17 @@ var (
 		Arity:    -2,
 		KeySpecs: KeySpecs{BeginIndex: 1, Step: 1},
 	}
+	expireatCmdMeta = DiceCmdMeta{
+		Name: "EXPIREAT",
+		Info: `EXPIREAT sets a expiry time(in unix-time-seconds) on the specified key in args
```
Comment (JyotinderSingh) [inline core/commands.go:575]:
```suggestion
		Info: `EXPIREAT sets an expiry time(in unix-time-seconds) on the specified key in args
```

**Comment by ashwaniYDV** [comment]
@JyotinderSingh I have covered all the edge cases as suggested by you in one of the PR.

File: core/commands.go [hunk 2]
```diff
@@ -570,6 +570,17 @@ var (
...
+		args should contain 2 values, key and the expiry time to be set for the key
+		The expiry time should be in integer format; if not, it returns encoded error response
+		Returns RespOne if expiry was set on the key successfully.
+		Once the time is lapsed, the key will be deleted automatically`,
+		Eval:     evalEXPIREAT,
+		Arity:    -3,
+		KeySpecs: KeySpecs{BeginIndex: 1, Step: 1},
+	}
```
Comment (JyotinderSingh) [inline core/commands.go:583]: nit: could you reposition this block to the end of the initialiser list

File: core/eval.go [hunk 3]
```diff
@@ -616,6 +616,35 @@ func evalEXPIRETIME(args []string) []byte {
...
+	}
+
+	obj := Get(key)
+
+	// 0 if the timeout was not set. e.g. key doesn't exist, or operation skipped due to the provided arguments
+	if obj == nil {
+		return RespZero
+	}
+
+	setUnixTimeExpiry(obj, exUnixTimeSec)
```
Comment (JyotinderSingh) [inline core/eval.go:642]: An expiry in the past should immediately delete the key.

File: core/eval.go [hunk 4]
```diff
@@ -616,6 +616,35 @@ func evalEXPIRETIME(args []string) []byte {
 	return Encode(int(exTimeMili/1000), false)
 }
 
+// evalEXPIREAT sets a expiry time(in unix-time-seconds) on the specified key in args
+// args should contain 2 values, key and the expiry time to be set for the key
+// The expiry time should be in integer format; if not, it returns encoded error response
+// Returns RespOne if expiry was set on the key successfully.
+// Once the time is lapsed, the key will be deleted automatically
+func evalEXPIREAT(args []string) []byte {
+	if len(args) <= 1 {
```
Comment (JyotinderSingh) [inline core/eval.go:625]: From redis docs,

> The EXPIREAT command supports a set of options:
NX -- Set expiry only when the key has no expiry
XX -- Set expiry only when the key has an existing expiry
GT -- Set expiry only when the new expiry is greater than current one
LT -- Set expiry only when the new expiry is less than current one
A non-volatile key is treated as an infinite TTL for the purpose of GT and LT. The GT, LT and NX options are mutually exclusive.

We can add support for these commands in a separate PR, could you help with creating an issue for the same?

File: core/store.go [hunk 5]
```diff
@@ -316,6 +316,13 @@ func setExpiry(obj *Obj, expDurationMs int64) {
 	expires[obj] = uint64(utils.GetCurrentTime().UnixMilli()) + uint64(expDurationMs)
 }
 
+// setUnixTimeExpiry sets the expiry time for an object.
+// This method is not thread-safe. It should be called within a lock.
+func setUnixTimeExpiry(obj *Obj, exUnixTimeSec int64) {
+	// convert unix-time-seconds to unix-time-milliseconds
+	expires[obj] = uint64(exUnixTimeSec * 1000)
```
Comment (JyotinderSingh) [inline core/store.go:323]: Acquire a lock on the expires map when modifying it.

**Review by JyotinderSingh — CHANGES_REQUESTED** [review]
Thanks for these changes @ashwaniYDV, have left a few comments.

File: core/store.go (hunk 5 above)
Comment (JyotinderSingh) [reply core/store.go:323]: Don't modify this method, define another method which calls this one while acquiring the respective lock.
Additionally, you can consider calling the delete API as a part of that new method (within the same critical section) to make the operation atomic.

The flow should be:
1. Acquire lock
2. Check if expiry in the past, delete if yes.
3. If not, set Expiry.
4. Release lock.

Comment (ashwaniYDV) [reply core/store.go:323]: @JyotinderSingh expire map is defined like this
```
expires   map[*Obj]uint64 // Does not need to be thread-safe as it is only accessed by a single thread.
```

in the codebase I can see that we do not take lock on it, while modifying it

File: core/eval.go (hunk 4 above)
Comment (ashwaniYDV) [reply core/eval.go:625]: sure I will create an issue for both `EXPIRE` and `EXPIREAT` commands

File: core/store.go (hunk 5 above)
Comment (JyotinderSingh) [reply core/store.go:323]: I guess you can avoid a lock here since it is only accessed by the eval loop (and the loop is single-threaded). We will anyway be moving away from the current architecture in some time.

Also you're right, we can depend on the db for clearing out the key, no need to eagerly delete.

**Review by JyotinderSingh — APPROVED** [review]
Thanks for adding the EXPIREAT command @ashwaniYDV! Approved.
